src/container/srv_target.c

```diff
@@ -524,6 +524,17 @@ ds_cont_local_open(uuid_t pool_uuid, uuid_t cont_hdl_uuid, uuid_t cont_uuid,
 		*cont_hdl = hdl;
 	}
 
+	if (resync) {
+		/* Sync-up the DTXs status. */
+		rc = dtx_resync(hdl->sch_pool->spc_hdl, pool_uuid, cont_uuid,
+				hdl->sch_pool->spc_map_version, false);
+		if (rc != 0)
+			/* Just warning, not fail out. */
+			D_WARN("Fail to resync some DTX(s) status, that may "
+			       "cause some potential unexpected IO result in "
+			       "subsequent operations: rc = %d\n", rc);
+	}
+
```

Comment:
Could you please move this outside this function, since ds_cont_local_open() suppose to be general. thanks.

Reply:
maybe cont_open_one()? if it has to be in open()? does it?

Reply:
We only need to trigger DTX-resync when open the container for the first time. If move it to cont_open_one(), then may trigger DTX-resync multiple times if found cached compatible container handle.

Reply:
Could you please add some comments why do we need dtx_resync() during container open? Thanks.  Can the dtx_resync to be asynchronous? a bit concern about the open rpc handling time.

Reply:
I will add comment in the patch.
We need re-sync the DTX status during container open, because if a server crashed, then former cached committable DTX on such server be lost. If such server is not evicted and start again, then before accessing the objects, we need to re-sync former related committable DTX status.

Reply:
I don't quite see why this re-sync job is done in container open, it looks to me it should be done:
- when starting target, or;
- preceding rebuild scan; (where I suppose no cont open being involved at all)

Probably we'd make ds_cont_local_open() an pure local VOS operation to avoid blocking on network partition between servers. Imaging an application does cont open/close repeatedly, will the dtx_resync() being called every time on cont open?

Reply:
According to our current DTX commit mechanism, there will be very limited DTXs (we can set/adjust the threshold) to be re-synced, even if it is handled when open the container, it can be handled very quickly (within 1 sec?). On the other hand, once the container is opened, it will be cached on server and shared by multiple clients. Even if some client open/close the container repeatedly, as long as at least one client still holds the handle, the container will not be really closed, then the dtx_resync() will NOT be re-triggered when others open the container.

Reply:
It is not impossible to re-sync DTX status when server start, but there are something to be considered: there may be multiple pools on the server, and the pool may contain many containers. If we trigger dtx_resync() during the server start, then means that all the DTXs belong to all the containers will be handled at that time. Then the total time for such process will be multiple times than single container case. One the other hand, it is possible that several servers start at almost the same time, then it is possible that when server_A is ready for dtx_resync(), the server_B may be not ready for the dtx_resync() yet, then server_A's dtx_resync() will either be booked or failed. That may cause the server_A to be unavailable because of server_B's absence or delay. It is possible that server_A only has one container that has shard on the server_B. But other un-related containers on the server_A are also unavailable because of un-related server_B. As my understand, such case should be avoided as much as possible.

Reply:
So, trigger the dtx_resync() when server start may not bring too much value than current solution.

Reply:
If you have to re-create the dtx table in DRAM during open? could you do it in an asynchronous ULT to avoid delay open?  So it only check the PMEM before resync finish. it should be fine?

Reply:
The async mode DTX resync may cause correctness issue. Because after opening the container, the application may access some data which DTX status was 'committable' before the container close/crash. But before the DTX resync done, it is invisible to application. That is not expected.

Reply:
I thought you can check the status from PMEM correctly before dtx resync finish?

Reply:
No, even if it is on the leader, before dtx resync done, the visibility is still unknown. There is only two status on-disk, either 'committed' or 'prepared'. The 'committable' status is only temporary in DRAM.

Reply:
Hmm, when opening the container, it can start an asynchronous ULT to resync the dtx. Then if there are others tries to access the container, it will check if resync dtx finished,

if it is not finished, then it will check dtx in pmem directly to see if there are related data uncommitted for this access. if it does, it will wait simply.

otherwise, it will access the data as normal?

Do I miss sth?

Reply:
I will update the patch to resync DTX when open container via asynchronous ULT.